<a href="https://colab.research.google.com/github/qkrwogus1213/AI18_Section1_Project/blob/main/n212a_generalization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="right" src="https://ds-cs-images.s3.ap-northeast-2.amazonaws.com/Codestates_Fulllogo_Color.png" width=100>

## *AIB / SECTION 2 / SPRINT 1 / NOTE 2*

# 📝 Assignment
---

# 🌱 **기본 과제**

이번 과제에서도 n211 과제에서 다루었던 housing price dataset을 가지고 배운 내용을 연습해 보겠습니다. 

과제에서 사용
할 데이터셋은 다음 링크에서 다운로드 받을 수 있습니다.

   - [House Sales in King County, USA](https://www.kaggle.com/harlfoxem/housesalesprediction?select=kc_house_data.csv)

## **Part 1. Validation**

### **1-1. 3-way holdout Method**

- 데이터셋을 불러오세요. 
- 데이터를 특성과 타겟으로 나누어 주세요.
  - 특성은 `bedrooms`, `bathrooms`, `sqft_living`을 사용합니다.
  - 타겟은 `price`입니다. 
- `train_test_split()`을 사용하여 전체 데이터를 훈련과 테스트 데이터로 나누어주세요. 
  - 훈련과 테스트 데이터는 8:2의 비로 나눕니다.
  - `random_state=2`로 설정합니다. 
- `train_test_split()`을 사용하여 훈련 데이터를 훈련과 검증 데이터로 나누어주세요.
  - 훈련과 검증 데이터는 8:2의 비로 나눕니다.
  - `random_state=2`로 설정합니다.

##### **Q1. 데이터를 분리하고 난 후 검증 세트(X_val)의 개수를 제출하세요.**

In [120]:
# 데이터셋을 불러옵니다. 

import pandas as pd
import numpy as np

df = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kc_house_data/kc_house_data.csv')
df

id             date      price  bedrooms  bathrooms  \
0      7129300520  20141013T000000 221,900.00         3       1.00   
1      6414100192  20141209T000000 538,000.00         3       2.25   
2      5631500400  20150225T000000 180,000.00         2       1.00   
3      2487200875  20141209T000000 604,000.00         4       3.00   
4      1954400510  20150218T000000 510,000.00         3       2.00   
...           ...              ...        ...       ...        ...   
21608   263000018  20140521T000000 360,000.00         3       2.50   
21609  6600060120  20150223T000000 400,000.00         4       2.50   
21610  1523300141  20140623T000000 402,101.00         2       0.75   
21611   291310100  20150116T000000 400,000.00         3       2.50   
21612  1523300157  20141015T000000 325,000.00         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650    1.00           0     0  ...      7   
1             2570      7242    2.00           0     0  ...      7   
2              770     10000    1.00           0     0  ...      6   
3             1960      5000    1.00           0     0  ...      7   
4             1680      8080    1.00           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131    3.00           0     0  ...      8   
21609         2310      5813    2.00           0     0  ...      8   
21610         1020      1350    2.00           0     0  ...      7   
21611         1600      2388    2.00           0     0  ...      8   
21612         1020      1076    2.00           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode   lat  \
0            1180              0      1955             0    98178 47.51   
1            2170            400      1951          1991    98125 47.72   
2             770              0      1933             0    98028 47.74   
3            1050            910      1965             0    98136 47.52   
4            1680              0      1987             0    98074 47.62   
...           ...            ...       ...           ...      ...   ...   
21608        1530              0      2009             0    98103 47.70   
21609        2310              0      2014             0    98146 47.51   
21610        1020              0      2009             0    98144 47.59   
21611        1600              0      2004             0    98027 47.53   
21612        1020              0      2008             0    98144 47.59   

         long  sqft_living15  sqft_lot15  
0     -122.26           1340        5650  
1     -122.32           1690        7639  
2     -122.23           2720        8062  
3     -122.39           1360        5000  
4     -122.05           1800        7503  
...       ...            ...         ...  
21608 -122.35           1530        1509  
21609 -122.36           1830        7200  
21610 -122.30           1020        2007  
21611 -122.07           1410        1287  
21612 -122.30           1020        1357  

[21613 rows x 21 columns]

In [121]:
from sklearn.model_selection import train_test_split

feature = df[['bedrooms', 'bathrooms', 'sqft_living']]
target = df['price']


X_train, X_test, y_train, y_test  = train_test_split(feature, target, test_size = 0.2, random_state = 2)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state = 2)


print("validation set : ", X_val.shape)
print("X_train",X_train.shape)
print("X_test",X_test.shape)

validation set :  (3458, 3)
X_train (13832, 3)
X_test (4323, 3)


##### **Q2. 기준모델과 다중선형회귀 모델의 성능으로 옳은 것을 고르세요. (반올림하여 소수점 둘째 자리까지)**
- 모델을 만들기 전에 훈련 데이터 타겟의 평균값을 사용한 기준모델을 만들고, 기준모델의 RMSE, $R^2$ score를 확인해보세요.
- 다중선형회귀 모델을 만들고 훈련셋과 검증셋의 RMSE, $R^2$ score를 비교해보세요.
- 마지막으로 테스트셋을 사용해서 모델의 일반화 성능을 확인해보세요.
  - **퀴즈 보기**에 나온 다중선형회귀 모델의 score는 **테스트셋을 사용한 값**입니다.
  - 성능은 RMSE, $R^2$ score를 사용하세요.
- MSE, MAE도 확인해보세요.

In [122]:
import numpy as np
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures
from IPython.display import display, HTML
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt

pd.options.display.float_format = '{:,.2f}'.format



def eval_models(y_true, y_pred) :
    mse = mean_squared_error(y_true, y_pred).round(3)
    rmse = np.sqrt(mse).round(3)
    mae = mean_absolute_error(y_true, y_pred).round(3)
    r2 = r2_score(y_true, y_pred)

    return mse, rmse, mae, r2

model = LinearRegression()
model.fit(X_train, y_train)

y_train_pred = model.predict(X_train)
y_val_pred = model.predict(X_val)
y_test_pred = model.predict(X_test)

baseline = [y_train.mean()] * len(y_train)

comparison = pd.DataFrame(index=['mse','rmse', 'mae', 'r2'], columns=['Base', 'Train', 'Validation', 'Test'])
comparison['Base'] = eval_models(y_train, baseline)
comparison['Train'] = eval_models(y_train, y_train_pred)
comparison['Validation'] = eval_models(y_val, y_val_pred)
comparison['Test'] = eval_models(y_test, y_test_pred)

display(comparison)




Base             Train        Validation              Test
mse  135,310,752,597.86 66,385,260,945.86 65,548,488,927.30 67,507,634,927.19
rmse         367,846.10        257,653.37        256,024.39        259,822.31
mae          234,667.59        170,774.55        170,429.07        172,277.08
r2                 0.00              0.51              0.48              0.52

### **1-2. Cross_validation**

- **1-1**과 동일한 조건으로 데이터를 다시 train set과 test set으로만 나누세요.
  - K-Fold CV 방법을 사용하기 때문에 validation set을 따로 만들지 않습니다.
- 교차검증을 통해 나온 결과를 확인해보세요.
  - sklearn의 `cross_val_score` 라이브러리를 사용하세요.
  - **10-fold** cross validation
을 수행하세요.
  - cv 수행 시 scoring metric은 **RMSE** 로 연산하도록 설정해 주세요.
  


##### **Q3. cross_val_score 함수에 대한 설명으로 옳은 것을 고르세요.**

In [123]:
from sklearn.model_selection import KFold

target = 'price'
features = ['bedrooms', 'bathrooms', 'sqft_living']

def x_y_split(df):
  X = df[features]
  y = df[target]
  return X, y

df_train, df_test = train_test_split(df, test_size = 0.2, random_state = 2)
X_train, y_train = x_y_split(df_train)
X_test, y_test = x_y_split(df_test)

kf = KFold(n_splits=10)
print('k =',kf.get_n_splits())

cv_result = []
for train_idx, test_idx in kf.split(X_train) :
    X_train_cv, X_val_cv = X_train.iloc[train_idx], X_train.iloc[test_idx]
    y_train_cv, y_val_cv = y_train.iloc[train_idx], y_train.iloc[test_idx]

    model.fit(X_train_cv, y_train_cv)
    y_pred_cv = model.predict(X_val_cv)
    
    mse = mean_squared_error(y_val_cv, y_pred_cv).round(3)
    rmse = np.sqrt(mse).round(3)
    cv_result.append(rmse)

print('rmse for each fold : ', cv_result)
print('average mae for model : ', np.mean(cv_result).round(2))
print('std of result : ', np.std(cv_result).round(2))

k = 10
rmse for each fold :  [277394.946, 239026.738, 264116.472, 269107.239, 263092.248, 231691.037, 242326.233, 280749.301, 244181.512, 258555.786]
average mae for model :  257024.15
std of result :  15999.87


## **Part 2. Overfit/Underfit**



### **2-1. 과적합**

- 다항선형회귀를 만들고 최적의 `차수(degree)`를 찾아보세요.
  - 데이터를 **1-1**과 같이 다시 훈련/검증/테스트 셋으로 나누어 모델 학습 후 에러를 비교하세요.
  - 주어진 리스트 `degrees`의 요소 중에서 **검증셋의 에러가 가장 낮은** 차수를 찾습니다.
  - 차수를 찾기 위한 코드는 자유롭게 작성해보세요.
    - EX.1) for문을 통해 여러 개의 다항회귀 모델 결과값 출력
    - EX.2) 다항회귀모델의 차수를 직접 바꿔가면서 결과값 확인 등등
  - 에러를 비교하기 위한 평가지표는 자유롭게 사용해도 좋습니다.


##### **Q4. 다항선형회귀를 만들고 최적의 차수를 찾아 제출하세요.**

In [124]:
from matplotlib import pyplot as plt
from sklearn.preprocessing import PolynomialFeatures

feature = df[['bedrooms', 'bathrooms', 'sqft_living']]
target = df['price']

def PolynomialRegression(degree=3, **kwargs):
    return make_pipeline(PolynomialFeatures(degree), 
                         LinearRegression(**kwargs))


X_train, X_test, y_train, y_test  = train_test_split(feature, target, test_size = 0.2, random_state = 2)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state = 2)


degrees = [1, 2, 3, 4, 5]

best_degree = None

for i in degrees:
  poly_model = PolynomialRegression(i)
  poly_model.fit(X_train, y_train)
  y_pred_val = poly_model.predict(X_val)
  print("차수",i)
  print(eval_models(y_val, y_pred_val))



차수 1
(65548488927.295, 256024.391, 170429.067, 0.4819386780307512)
차수 2
(62903090387.428, 250804.885, 163571.373, 0.5028465385646457)
차수 3
(63818339336.948, 252622.919, 163982.178, 0.4956128846928465)
차수 4
(68310950009.708, 261363.636, 164179.109, 0.460105615763964)
차수 5
(73633408277.751, 271354.765, 166160.682, 0.4180396610548196)


##### **Q5. 차수가 5일 때 과적합이 발생했나요? 발생했다면 어떻게 해소할 수 있을까요?**

In [125]:
for i in degrees:
  poly_model = PolynomialRegression(i)
  poly_model.fit(X_train, y_train)
  y_predict_train = poly_model.predict(X_train)
  print("train 데이터 차수",i)
  print(eval_models(y_train, y_predict_train))

print('\n')
for i in degrees:
  poly_model = PolynomialRegression(i)
  poly_model.fit(X_train, y_train)
  y_predict_test = poly_model.predict(X_test)
  print("test 데이터 차수",i)
  print(eval_models(y_test, y_predict_test))  

train 데이터 차수 1
(66385260945.859, 257653.374, 170774.553, 0.5093866550047623)
train 데이터 차수 2
(58729419750.867, 242341.535, 163162.691, 0.5659663506165764)
train 데이터 차수 3
(58289838252.596, 241432.886, 163024.194, 0.5692150318176692)
train 데이터 차수 4
(57521565368.08, 239836.539, 162473.931, 0.5748928724161926)
train 데이터 차수 5
(58653631728.245, 242185.119, 164163.297, 0.5665264540907412)


test 데이터 차수 1
(67507634927.19, 259822.314, 172277.08, 0.5164493534799632)
test 데이터 차수 2
(71507842793.857, 267409.504, 166281.679, 0.4877962522088598)
test 데이터 차수 3
(79108328355.886, 281262.028, 167085.385, 0.4333547107247119)
test 데이터 차수 4
(135286338321.465, 367812.912, 169555.262, 0.030957069800640946)
test 데이터 차수 5
(294351602313.621, 542541.798, 174167.489, -1.108412000457041)


train set r2는 0.566
test set r2는 -1.108
으로 차수가 5일때 과적합이 발생함

## **Part 3. Generalization**

### **3-1. 일반화**

- **테스트셋**을 딱 한 번만 사용해서 모델의 **`일반화`** 성능을 확인하세요.
  - **다항선형회귀 모델**을 사용하고 **2-1**에서 구한 최적의 차수를 사용합니다. 
  - 학습시킨 다항선형회귀 모델의 **`훈련셋과 테스트셋의 score`**를 구해보세요.
  - 평가지표는 **RMSE**와 **$R^2$ score**를 사용하세요.

##### **Q6. 최종모델은 기준모델보다 개선된 모델인가요? (테스트셋에서의 RMSE와 $R^2$를 기준모델과 비교해보세요.)**


In [126]:
poly_model = PolynomialRegression(degree=2)
poly_model.fit(X_train,y_train)

y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)


display(comparison.iloc[:,0])

test_result = eval_models(y_test, y_pred_test)
train_result = eval_models(y_train, y_pred_train)
print("test", test_result)
print("train", train_result)


#r2는 증가 rmse는 감소

mse    135,310,752,597.86
rmse           367,846.10
mae            234,667.59
r2                   0.00
Name: Base, dtype: float64

test (67478861881.847, 259766.938, 171759.023, 0.5166554520152344)
train (66401350537.842, 257684.595, 170355.128, 0.5092677465538505)


# 🥇 **도전 과제**

- **과제에서 만든 모델의 성능을 개선해보세요.**
    - **`EDA`**를 통해 추가로 학습에 사용할 특성을 찾고 모델을 새로 학습해보세요.
    - 여러분의 다양한 아이디어로 **`특성 공학(Feature Engineering)`**을 수행해 볼 수도 있습니다.
    - 마지막에 모델 성능의 개선 여부와 왜 그 결과가 나왔는지에 대한 **`인사이트`**를 공유해주세요.
